processor/resourcedetectionprocessor/internal/resourcedetection.go

```diff
@@ -117,65 +118,94 @@ func NewResourceProvider(logger *zap.Logger, timeout time.Duration, attributesTo
 	}
 }
 
-func (p *ResourceProvider) Get(ctx context.Context, client *http.Client) (resource pcommon.Resource, schemaURL string, err error) {
-	p.once.Do(func() {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, client.Timeout)
-		defer cancel()
-		p.detectResource(ctx, client.Timeout)
-	})
+func (p *ResourceProvider) Get(ctx context.Context, client *http.Client) (pcommon.Resource, string, error) {
+	p.mu.RLock()
+	res := p.detectedResource
+	p.mu.RUnlock()
 
-	return p.detectedResource.resource, p.detectedResource.schemaURL, p.detectedResource.err
+	if res != nil {
+		return res.resource, res.schemaURL, res.err
+	}
+
+	// First access: ensure only one goroutine computes.
+	p.initOnce.Do(func() { _, _, _ = p.Refresh(ctx, client) })
+
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	res = p.detectedResource
+	if res == nil {
+		return pcommon.NewResource(), "", nil
+	}
+	return res.resource, res.schemaURL, res.err
 }
 
-func (p *ResourceProvider) detectResource(ctx context.Context, timeout time.Duration) {
-	p.detectedResource = &resourceResult{}
+// Refresh recomputes the resource, replacing any previous result.
+func (p *ResourceProvider) Refresh(ctx context.Context, client *http.Client) (pcommon.Resource, string, error) {
+	ctx, cancel := context.WithTimeout(ctx, client.Timeout)
+	defer cancel()
 
+	res, schemaURL, err := p.detectResource(ctx, client.Timeout)
```

Comment:
Why does this need to take `client.Timeout` when it's also taking a context that just had its timeout set from that value?